### PR TITLE
Support fetching PSI stats for cgroupv2 containers

### PR DESCRIPTION
**Pull request by dqminh**
This supports fetching PSI stats for cgroupv2 containers. We read the PSI metrics if they are available from:
- cpu.pressure
- memory.pressure
- io.pressure

See more about PSI at https://www.kernel.org/doc/html/latest/accounting/psi.html

### CODE REVIEW

File: libcontainer/cgroups/stats.go
Comment (AkihiroSuda) [inline libcontainer/cgroups/stats.go:38]: Maybe: `Avg map[int]float64` (key=10,60,300)

File: libcontainer/cgroups/fs2/psi_test.go
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/psi_test.go:15]: Can we have integration tests too?

Comment (dqminh) [reply libcontainer/cgroups/fs2/psi_test.go:15]: added one for runc events

File: libcontainer/cgroups/stats.go
Comment (dqminh) [reply libcontainer/cgroups/stats.go:38]: i don't like that very much because the key avg10,60,300 is fixed, so invidual fields are much better for discovery.

Comment (kolyshkin) [reply libcontainer/cgroups/stats.go:38]: I was thinking about it, too. If this is fixed in the same way as, say, loadavg is fixed for 1, 5, 15 minutes, we can just use an array with 3 numbers.

OTOH maybe the best way is to have
```
Some, Full map [string]float64
```
where map keys are `avgXXX` and `total`.

In case we want `total` as an integer, maybe it's better to use whatever @AkihiroSuda is proposing, plus a separate Total field.

**Comment by kolyshkin** [comment]
Can we please have `s/Psi/PSI/` since this is an acronym? I know we use things like `Json` before, but in the new code we can do better.

File: tests/integration/events.bats
Comment (kolyshkin) [inline tests/integration/events.bats:33]: Can you move this to helpers.bash, function `requires`?

File: libcontainer/cgroups/stats.go
Comment (dqminh) [reply libcontainer/cgroups/stats.go:38]: the drawbacks of array or map is non-discoverability of keys and extra allocation where it's not necessary. I don't yet see anything that is bad with explicit keys.

We can make the same arguments with a lot of other statistics too such as io.stat. I think explicit struct are much better to use.

File: tests/integration/events.bats
Comment (dqminh) [reply tests/integration/events.bats]: done, this is now `requires psi`.

**Review by kolyshkin — APPROVED** [review]
LGTM

File: tests/integration/helpers.bash
Comment (kolyshkin) [inline tests/integration/helpers.bash:454]: You don't need to `cat` the file, mere existence check should be sufficient.

IOW
```bash
if [ ! -f /sys/fs/cgroup/cpu.pressure ]; then
...
```

Comment (kolyshkin) [reply tests/integration/helpers.bash:454]: Ah, I see, you *do* need to actually cat the file. I should have read better before writing comments 🤦🏻

File: tests/integration/helpers.bash
Comment (kolyshkin) [inline tests/integration/helpers.bash:453]: typo: `s/PIS/PSI/`.

Comment (kolyshkin) [reply tests/integration/helpers.bash:453]: nit: period missing at EOL.

Comment (kolyshkin) [reply tests/integration/helpers.bash:453]: typo: `s/enable/enabled/`.

Comment (kolyshkin) [reply tests/integration/helpers.bash:453]: typo: `s/ENOSUPP/ENOTSUP/`.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
a few typos in helpers.bash, otherwise LGTM

File: tests/integration/helpers.bash
Comment (dqminh) [reply tests/integration/helpers.bash]: Thanks, updated.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by dqminh** [comment]
ping @opencontainers/runc-maintainers

**Comment by dqminh** [comment]
ping @opencontainers/runc-maintainers again

**Comment by kolyshkin** [comment]
@dqminh can you please rebase? We have changed something in the tests and they need to be re-run.

**Comment by dqminh** [comment]
@kolyshkin thanks, rebase is done. I had to add `# shellcheck disable=SC2030` for the integration test.

**Comment by bobbypage** [comment]
ping @kolyshkin, this would be great to get in, especially for use downstream in cAdvisor / k8s

File: libcontainer/cgroups/fs2/fs2.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/fs2.go:117]: nit: extra space after `(`

File: libcontainer/cgroups/fs2/psi.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/psi.go:45]: I think it'll be more readable to use
```go
          for _, f := range psi {
```
(and use `f` instead of `psi[i]` below).

File: libcontainer/cgroups/fs2/psi_test.go
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/psi_test.go:15]: nit: capitalize `PSI`

File: libcontainer/cgroups/fs2/psi_test.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/psi_test.go:12]: nit: capitalize `PSI`

**Review by kolyshkin — COMMENTED** [review]
A few nits, otherwise looks good.

**Comment by dqminh** [comment]
@kolyshkin updated the nits, also added a fix to check for ENOTSUP when kernel compiled with PSI_DEFAULT_DISABLED, so reading *.pressure file return ENOTSUP as seen in https://cirrus-ci.com/task/4929935776153600

**Comment by bobbypage** [comment]
ping @kolyshkin for another round of a reviews :)

File: libcontainer/cgroups/fs2/fs2.go
Comment (kolyshkin) [inline libcontainer/cgroups/fs2/fs2.go:122]: It's probably better to use `unix.ENOTSUP` from `golang.org/x/sys/unix` here (as `syscall` is considered an obsoleted package).

File: tests/integration/helpers.bash
Comment (kolyshkin) [inline tests/integration/helpers.bash:456]: No need to use `[[` here.

```bash
if ! cat /sys/fs/cgroup/cpu.pressure &>/dev/null; then
```

should be sufficient.

**Comment by bobbypage** [comment]
gentle ping to @dqminh regarding the updates mentioned. would be great to get this in so we can consume PSI metrics downstream in cAdvisor / k8s

**Comment by AkihiroSuda** [comment]
@kolyshkin Maybe we can address those issues after merging?

**Comment by Furisto** [comment]
Hey @dqminh this is a great PR. There has been no activity for some time so I am wondering if you are still interested? Otherwise I would be willing to take over and make the requested changes.

**Comment by Furisto** [comment]
@kolyshkin Given that this PR has stalled, how would you like to proceed? I can make the requested changes. Would be great if we could make PSI metrics available.

**Comment by dqminh** [comment]
@thaJeztah i'm really sorry about the delay. We have been running with the patch for so long (!) that i forgot it hasn't been merged.

I left a few comments in the PR on my fork. Generally i think we can try to make the refactoring of existing practices separately from the current change.

**Comment by szuecs** [comment]
@dqminh I read the comments and tried to fix everything and opened https://github.com/dqminh/runc/pull/37 . I have here issue I can't compile it and test wait for fetching containers since 15 minutes.
Dirty, because I did not commit and pushed changes when I tried to `make`.

```
% make
go build -trimpath "-buildmode=pie"  -tags "seccomp" -ldflags "-X main.gitCommit=v1.1.0-181-g6414629a-dirty -X main.version=1.1.0+dev " -o runc .
# pkg-config --cflags  -- libseccomp
Package libseccomp was not found in the pkg-config search path.
Perhaps you should add the directory containing `libseccomp.pc'
to the PKG_CONFIG_PATH environment variable
Package 'libseccomp', required by 'virtual:world', not found
pkg-config: exit status 1
make: *** [Makefile:33: runc] Error 2
zsh: exit 2     make

..

[~/go/src/github.com/opencontainers/runc]% make test 
docker build  -t runc_dev:psi .
Sending build context to Docker daemon  33.68MB
Step 1/17 : ARG GO_VERSION=1.18
Step 2/17 : ARG BATS_VERSION=v1.3.0
Step 3/17 : ARG LIBSECCOMP_VERSION=2.5.3
Step 4/17 : FROM golang:${GO_VERSION}-bullseye
1.18-bullseye: Pulling from library/golang
17c9e6141fdb: Downloading [=======================>                           ]  26.28MB/55.05MB
de4a4c6caea8: Download complete
4edced8587e6: Download complete
a7969cffbf46: Downloading [=====================>                             ]  23.61MB/54.59MB
efa7548a174d: Downloading [=========>                                         ]  16.11MB/85.97MB
ca7db62c462d: Waiting
623d7734067d: Waiting

```
I hope this helps to get it merged. :)

**Comment by h-vetinari** [comment]
Stray comment: As of linux 6.1 there's an [interface](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=bd9a3dba185ce6701b41f0341470d3f53bbbbaed) called `cgroup.pressure` to switch PSI accounting on/off per cgroup. I suppose the code here will have to check if accounting is turned off.

**Comment by szuecs** [comment]
FYI: I created https://github.com/opencontainers/runc/pull/3679 as follow up

**Pull request closed without merging**
